File: JavaScript/Board.js


function Board(width, height, numberOfApples) {
	this.width = width;
	this.height = height;
	this.board = new Array();
	
	this.addApple = function() {
		var added = false;
		while (!added) {
			var x = Math.round(Math.random() * (this.width - 1));
			var y = Math.round(Math.random() * (this.height - 1));
			if (this.board[x] == null
				|| this.board[x][y] == null) {
				if (this.board[x] == null) {
					this.board[x] = new Array();
				}
				this.board[x][y] = new Apple(x, y);
				added = true;
				if (this.view != null)
					this.view.addApple(this.board[x][y]);
			}
		}
	}
	
	this.setView = function(view) {
		this.view = view;
		view.setBoardData(this.board);
		view.updateCompletely();
	}
	
	this.getWidth = function() {
		return this.width;
	}
	
	this.getHeight = function() {
		return this.height;
	}
	
	this.isApple = function(x, y) {
		x %= this.width;
		y %= this.height;
		if (this.board[x] && this.board[x][y]) {
			return (this.board[x][y] instanceof Apple);
		}
		return false;
	}
	
	this.isSnake = function(x, y) {
		x %= this.width;
		y %= this.height;
		if (this.board[x] && this.board[x][y]) {
			return (this.board[x][y] instanceof SnakeElement);
		}
		return false;
	}
	
	this.remove = function(element) {
		this.board[element.x][element.y] = null;
		this.view.remove(element);
	}
	
	this.add = function(element) {
		if (this.board[element.x] == null) {
			this.board[element.x] = new Array();
		}
		this.board[element.x][element.y] = element;
		this.view.add(element);
	}
	
	while (numberOfApples >= 0) {
		this.addApple();
		numberOfApples--;
	}
}

File: Rust/src/board.rs
extern crate rand;

use std::cell::RefCell;
use std::rc::Rc;

use self::rand::Rng;
use game_elements::{GameElement, new_apple, SnakeElement};
use board_view::BoardViewer;

pub struct Board {
    width:  usize,
    height: usize,
    board:  Vec<Vec<GameElement>>,
    random: rand::ThreadRng
}

impl Board {
    pub fn add_apple(&mut self) {
        let mut added = false;

        while !added {
            let x = self.random.gen_range(0, self.width);
            let y = self.random.gen_range(0, self.height);

            added = match self.board[x][y] {
                GameElement::Empty => {
                    self.board[x][y] = new_apple(x, y);
                    self.view(new_apple(x, y));
                    true
                },
                _ => false
            }
        }
    }

    pub fn get_width(&self) -> usize {
        self.width
    }

    pub fn get_height(&self) -> usize {
        self.height
    }

    pub fn add(&mut self, elem: Rc<RefCell<SnakeElement>>) {
        let x;
        let y;
        {   let el = elem.borrow();
            x = el.get_x();
            y = el.get_y();
        }
        self.board[x][y] = GameElement::SnakeElement(elem.clone());
        self.view(GameElement::SnakeElement(elem.clone()))
    }

    pub fn is_apple(&self, x: usize, y: usize) -> bool {
        let _x = x % self.width;
        let _y = y % self.height;

        match self.board[_x][_y] {
            GameElement::Apple(_) => true,
            _                     => false
        }
    }

    pub fn is_snake(&self, x: usize, y: usize) -> bool {
        let _x = x % self.width;
        let _y = y % self.height;

        match self.board[_x][_y] {
            GameElement::SnakeElement(_) => true,
            _                            => false
        }
    }

    pub fn remove(&mut self, elem: Rc<RefCell<SnakeElement>>) {
        let e = elem.borrow();
        self.board[e.get_x()][e.get_y()] = GameElement::Empty;
        self.hide(e.get_x(), e.get_y());
    }

    pub fn for_each_element<F>(&self, f:F) where F: Fn(GameElement) {
        for row in self.board.iter() {
            for e in row.iter() {
                f(e.clone())
            }
        }
    }
}

pub fn new(width: usize, height: usize, num_of_apples: usize) -> Board {
    let mut b = Board {width:  width,
                       height: height,
                       board:  vec![vec![GameElement::Empty; height]; width],
                       random: rand::thread_rng()};

    for _ in 0..num_of_apples {
        b.add_apple();
    }
    b
}
